documentation/api/query/v4/pql.markdown

```diff
@@ -0,0 +1,361 @@
+---
+title: "PuppetDB 3.2 » API » v4 » Puppet query language (PQL)"
+layout: default
+canonical: "/puppetdb/latest/api/query/v4/pql.html"
+---
+
+[entities]: ./entities.html
+[subquery]: #subqueries
+[ast]: ./ast.html
+
+> **Experimental Feature**: This featureset is experimental, and it may be altered or removed in
+> future release.
+
+Puppet Query Language (PQL) is a query language designed with PuppetDB and Puppet data in mind. It
+provides a string based query language as an alternative to the [AST query language][ast] PuppetDB has
+always supported.
+
+## Query Structure
+
+A PQL query has the following structure:
+
+    <entity> [<projection>] { <filter> <modifiers> }
+
+Which is broken up into the following parts:
+
+* `entity`: Rquired. The entity context that this query executes on.
+* `projection`: Optional. Modifies the output of how the `entity` fields are returned.
+* `filter`: Optional. The filter to match on for this `entity`.
+* `modifiers`: Optional. Contains modifiers for the query.
+
+As an example, if you wanted to query for `nodes` related data, but only see the `certnames` field
+for each node, with a regex filter across certname:
+
+    nodes[certname] { certname ~ '^web' }
+
+In this case, this would return only the certname field of nodes starting with `web`.
+
+## Entity
+
+The entity or context of a query (or subquery) defines what results you will get returned when
+performing a query, and provides the main context for any projections or filters in the query.
+There are many entities, for a full list see the [entities] documentation.
+
+For PQL queries, the entity context is the minimal amount of information one must provide, as it
+defines the results returned. For example, if you wanted to see all node information, you could
+provide a query as follows:
+
+    nodes {}
+
+And it would be enough to return all node data, without filtering or pagination.
+
+The entity context can also be used within a subquery, see the [subquery] section for more details.
+
+## Projection
```

Comment:
@mullr I think you make some good points, I'm moving this conversation into another ticket for now: https://tickets.puppetlabs.com/browse/PDB-2358 ... so we can debate it on its own. This shouldn't be hard to change either way, if we decide its a good idea. I think this will pan out to be a team discussion, then execution on the details we decide. Thanks mate.